### ISSUE TITLE
[Bug]: useControlField's setValue not working as expected in certain cases

### ISSUE THREAD

**Issue by danecando** [airjp73/remix-validated-form#123]
### Which packages are impacted?

- [X] `remix-validated-form`
- [ ] `@remix-validated-form/with-zod`
- [ ] `@remix-validated-form/with-yup`
- [ ] `zod-form-data`

### What version of these packages are you using?

- `remix-validated-form`: ^4.5.0

### Please provide a link to a minimal reproduction of the issue.

https://stackblitz.com/edit/node-xe3wxt?file=app/routes/index.tsx

### Steps to Reproduce the Bug or Issue

Following the 'occasional field tracking' example pretty much 1:1 from the useControlField docs: https://www.remix-validated-form.io/reference/use-control-field

### Expected behavior

I expected for `value` to be updated after use `setValue` in the `useEffect` hook

``` javascript
 React.useEffect(() => {
    setValue('other value');
  }, [setValue]);
```

The value never gets set and stays `undefined`

### Platform

- OS: macOS
- Browser: Chrome
- Version: 102.0.5005.61 

### Additional context

I'm using useControlField successfully for other fields in the same form (in my actual app). The difference is the controlled field value is just wrapping a regular `useField` input ie: 

```javascript
export function InputFieldControlled({ name, ...props }: InputFieldProps) {
  const [value, setValue] =
    useControlField<JSX.IntrinsicElements['input']['value']>(name);

  return (
    <InputField
      name={name}
      value={value}
      onChange={(e) => {
        setValue(e.target.value);
      }}
      {...props}
    />
  );
}
```

Hopefully I'm missing something obvious before I spend time debugging tomorrow.

**Comment by airjp73** [comment]
Yeah, this is definitely a bug.

Specifically, this happens when the field is mounted at the same time as the form. Here's another stackblitz to demonstrate. If you click "Hide Some Field" and to hide the field, then "Show Some Field" to show the field, the value correctly populates.

Stackblitz: https://stackblitz.com/edit/node-ptizfx?file=app/routes/index.tsx

This should actually already be fixed in my in-progress `fieldarray` branch. That work is almost done, but I'm not sure when I'll have time to fully finish it up (I'm moving this month 😵). In the meantime, if you're open to adding a failing test [here](https://github.com/airjp73/remix-validated-form/blob/main/apps/test-app/cypress/integration/controlled-fields.ts) (will need to add a route [in this folder](https://github.com/airjp73/remix-validated-form/tree/main/apps/test-app/app/routes)), that would help a lot.

## Workarounds

Since this only happens on initial mount, can you set the `defaultValues`?

```ts
<ValidatedForm
  id="someForm"
  method="post"
  validator={validator}
  defaultValues={{
    someField: 'some-field',
    otherField: 'other-field',
  }}
>
```

If that really doesn't cut it, a dirtier workaround would be to make sure the component containing your effect doesn't mount until after the form itself does.

**Comment by danecando** [comment]
Good luck on the move that's always fun 😄 

I just hopefully tried initializing a zero-value and keeping the useEffect setting the actual value in the component (no go). I then reached for a tried and true hacky work around, the old `setTimeout, 0` and that works for now 🥴. 

I plan on taking some time over the next week to make some small initial contributions. I will def add a test case for this use case at the minimum. If there's anything in specific I can do to help get a fix for this one across the finish line let me know, if not I can workaround it for now anyways.

**Comment by DHFW** [comment]
@danecando @airjp73 I have a similar issue. I use the Dropdown component from the [Primefaces library](https://www.primefaces.org/primereact/dropdown/). When onChange the value should be updated, it always gives an empty string. When I replace the `<Dropdown>` component with a regular `<select> it works.  The e.value is logging the correct value from the select, so I'm lost what the problem is... @danecando Can you tell me your workaround in more details please? The code:
```
export const Select = ({ name, label, ...rest }: SelectProps) => {
  const { error, validate } = useField(name);
  const [value, setValue] = useControlField<string>(name);

  return (
    <>
      <Label name={name} label={label}></Label>
      <div className="flex flex-col gap-2">
        {/* <select
          id={name}
          name={name}
          value={value}
          onChange={(e) => {
            setValue(e.target.value);
          }}
        >
          {rest.options?.map((o) => (
            <option key={o.value} value={o.value}>
              {o.label}
            </option>
          ))}
        </select> */}
        <Dropdown
          id={name}
          inputId={name}
          name={name}
          value={value}
          onChange={(e) => {
            setValue(e.value);
            // validate();
          }}
          options={rest.options}
          appendTo={'self'}
        />
        {error && <span className="text-sm text-red-700">{error}</span>}
      </div>
    </>
  );
};
```

**Comment by airjp73** [comment]
@DHFW, could you provide a reproduction in a codesandbox so I can take a look at what's going on there?

**Comment by DHFW** [comment]
@airjp73 Is the issue clear to you or do you need more info? Please let me know. Thank you.

**Comment by airjp73** [comment]
@DHFW, sorry for the slow response time. It looks like something's going on with how primereact renders its select. When I inspect the element, it's rendering an empty option, and I'm not sure it's setting the selected value correctly. I don't have time to dig much deeper than that.

<img width="407" alt="Screen Shot 2022-07-25 at 10 02 38 PM" src="https://user-images.githubusercontent.com/25882770/180906600-4359d4e9-6b9e-4223-ab7c-5b101b86e579.png">
.

As a workaround, you can skip the `name` prop on the dropdown entirely and render your own hidden input. That has the expected result.

```jsx
return (
    <>
      <input type="hidden" value={value} name={name} />
      <Dropdown
        id={name}
        value={value}
        onChange={(e) => {
          console.log('Selected value:', e.value);
          setValue(e.value); // this does not seem to work
        }}
        options={props.options}
      />
    </>
  );
```

**Comment by airjp73** [comment]
About the original issue, sorry for the delays. We're finally all settled from moving and I'll be trying to make time to follow up on issues this week.

**Comment by DHFW** [comment]
@airjp73 Thank you so much for the workaround! I had similar problems with another component from PrimeReact (InputNumber) so I applied your suggestion and that worked! I can remove a lot of ugly workaround code of my own. Indeed it is something weird about PrimeReact that is causing this. For the Select/Dropdown functionality I am using React-select for now. It has a better UX than PrimeReact so that was not a real problem anymore. But still good to know that there is a solution!

Moving is hard work, so I understand that completely. Glad you could find some time to get back on this. I really love this library so far! I wish you the best!

**Comment by airjp73** [comment]
I still have documentation to write for new features, but I've released a beta with the fix (4.6.0-beta.0) if anyone wants to try it.

**Comment by ZipBrandon** [comment]
After running into this bug, I am using the `4.6.0-beta.0` branch and it is working for my cases that were failing in `4.5.5` version.

**Comment by airjp73** [comment]
Fix released in 4.6.0

**Comment by AnthonyLzq** [comment]
This issue still exists with MUI - DatePicker. I will try to provide an example as soon as I have free time, but if someone has the same issue, this solution worked for me:

> @DHFW, sorry for the slow response time. It looks like something's going on with how primereact renders its select. When I inspect the element, it's rendering an empty option, and I'm not sure it's setting the selected value correctly. I don't have time to dig much deeper than that.
> 
> <img alt="Screen Shot 2022-07-25 at 10 02 38 PM" width="407" src="https://user-images.githubusercontent.com/25882770/180906600-4359d4e9-6b9e-4223-ab7c-5b101b86e579.png">
> 
> .
> As a workaround, you can skip the `name` prop on the dropdown entirely and render your own hidden input. That has the expected result.
> 
> ```js
> return (
>     <>
>       <input type="hidden" value={value} name={name} />
>       <Dropdown
>         id={name}
>         value={value}
>         onChange={(e) => {
>           console.log('Selected value:', e.value);
>           setValue(e.value); // this does not seem to work
>         }}
>         options={props.options}
>       />
>     </>
>   );
> ```
